##################################################
# Prepare data with start population frequencies #
##################################################

suppressPackageStartupMessages({
  library(tidyverse)
  library(valr)
  library(optparse)
})

option_list = list(
  make_option(c("--outdir"),
              action = "store",
              default = NA,
              type = 'character',
              help = "The directory to store output files.")
)

opt <- parse_args(OptionParser(option_list=option_list))

if(!dir.exists(opt$outdir)){
  dir.create(opt$outdir)
  setwd(opt$outdir)
} else {
  setwd(opt$outdir)
}


#
# Get MAGIC line mosaics ----
#
# read filtered mosaics (generated by getMagicMosaics.R script)
mosaic_filtered <- read_csv("./data/external/founder_genotypes/magic_mosaics.csv")

# get all intervals
mosaic_intervals <- mosaic_filtered %>%
  distinct(chrom, start, end)

#
# Get parent genotypes ----
#
# accessions genotypes obtained with the getAccessionGenotypes.R"
accessions_genotypes <- read_csv("./data/external/founder_genotypes/accessions_genotypes.csv")

# Intersect the founder genotypes with the mosaic intervals
# this keeps only SNPs falling in those intervals
accessions_genotypes <- bed_intersect(accessions_genotypes, mosaic_intervals,
                                      suffix = c("", ".y")) %>%
  select(-matches(".y"), -.overlap)

# Make table of imputable SNPs
snps <- accessions_genotypes %>%
  mutate(cm = 0) %>%
  distinct(chrom, snp, cm, start) %>%
  select(chrom, snp, cm, start)

#
# Imputation ----
#
# Split mosaic by magic line otherwise RAM explodes
mosaic_genotypes <- mosaic_filtered %>%
  select(magic, chrom, start, end, accession) %>%
  split(.$magic)

# initate file headers for two files:
# SNP alleles
# Accession alleles
write_lines(paste("magic",
                  paste(snps$snp, collapse = ","),
                  sep = ","),
            "./data/external/founder_genotypes/imputed_magic_accession.csv")
write_lines(paste("magic",
                  paste(snps$snp, collapse = ","),
                  sep = ","),
            "./data/external/founder_genotypes/imputed_magic_snp.csv")

# impute each MAGIC line's genotypes - files are written on the go
walk(mosaic_genotypes, function(x){

  # Which MAGIC line is being processed at the moment
  magic_id <- unique(x$magic)

  # Group by accession - to do the intersect by accession
  x <- x %>% group_by(accession)

  # Intersect founder genotypes with MAGIC mosaic intervals
  magic_imputed <- accessions_genotypes %>%
    select(-major_allele_count, -ref_allele_count) %>%
    group_by(accession) %>%
    bed_intersect(x, suffix = c("", ".y"))

  # To ensure SNPs are in the right order, join with SNP table
  magic_imputed <- left_join(snps, magic_imputed, by = c("chrom", "snp", "start"))

  # # Write genotype table
  # magic_imputed %>%
  #   select(chrom, start, snp_allele, accession) %>%
  #   write_delim(paste0("./data_processed/startpop/startpop_magics/", magic_id, ".txt"),
  #               col_names = FALSE)

  # Write as a matrix
  write_lines(paste(magic_id,
                    paste(magic_imputed$accession, collapse = ","),
                    sep = ","),
              "./data/external/founder_genotypes/imputed_magic_accession.csv", append = TRUE)
  
  write_lines(paste(magic_id,
                  paste(magic_imputed$snp_allele, collapse = ","),
                  sep = ","),
            "./data/external/founder_genotypes/imputed_magic_snp.csv", append = TRUE)

  message("Processed ", magic_id)
  return(NULL)
})
